base/arrayshow.jl

```diff
@@ -483,13 +483,45 @@ show(io::IO, X::AbstractVector) = show_vector(io, X)
 
 ## Logic for displaying type information
 
+# given type `typeinfo` extracted from context, assuming a collection
+# is being displayed, deduce the elements type; in spirit this is
+# similar to `eltype`, but in some cases this would lead to incomplete
+# information: assume we are at the top level, and no typeinfo is set,
+# and that it is deduced to be typeinfo=Any by default, and consider
+# printing X = Any[1]; to know if the eltype of X is already diplayed,
+# we would compare eltype(X) to eltype(typeinfo) == Any, and deduce
+# that we don't need to print X's eltype because it's already known by
+# the context, which is wrong; even if default value of typeinfo is
+# not set to Any, then the problem would be similar one layer below
+# when printing an array like Any[Any[1]]; hence we must treat Any
+# specially (alternatively we could also specialize
+# `collection_eltype` for all known collection types, but this seems
+# like a heavy solution when the simpler solution seems to be enough)
+function collection_eltype(typeinfo::Union{DataType,UnionAll})::Union{DataType,UnionAll,Void}
+    if typeinfo == Any
+        # the current context knows nothing about what is being displayed, not even
+        # whether it's a collection or scalar
+        nothing
+    else
+        # this function is assumed to be called (e.g. from
+        # typeinfo_prefix) when the current object to display is a
+        # collection X, let's say of type Set{T}, but we also know
+        # that typeof(X) must be a sub-type of typeinfo != Any, for
+        # example typeinfo == AbstractSet{T}; then it's fair to assume
+        # that typeinfo is refering to a collection-like type (being
+        # a strict subtype of Any), and hence that what the context
+        # knows about eltype(X) is eltype(typeinfo)
+        eltype(typeinfo)
+    end
+end
+
 # X not constrained, can be any iterable (cf. show_vector)
 function typeinfo_prefix(io::IO, X)
     typeinfo = get(io, :typeinfo, Any)::Union{DataType,UnionAll}
     @assert X isa typeinfo
     # what the context already knows about the eltype of X:
     # (could use a more specific tool than eltype, maybe with _isleaftype)
-    eltype_ctx = eltype(typeinfo)
+    eltype_ctx = collection_eltype(typeinfo)
```

Comment:
I don't think special-casing `Any` in `collection_eltype` is the appropriate approach here. What happens with `Vector{Real}`? As you note, I think `isconcrete` would be better. Since you already `@assert X isa typeinfo`, if `isconcrete(eltype(typeinfo))` then you are sure that `eltype(typeinfo) === eltype(X)`.

Reply:
Sorry I had missed that comment. I agree that special-casing `Any` seems ad-hoc and brittle, but I don't see right away an approach with `isconcrete` which would be enough. Let's consider `Xs = Array{Real}[Real[1]]; X = Xs[1]`. When printing the elements of `Xs`, we will have `typeinfo == eltype(Xs) == Array{Real}`, which is not `isconcrete`; and eltype(X) == eltype(typeinfo) == Real`. So when printing `X`, we can omit the `Real` type specification, despite `eltype(typeinfo)` not being concrete. I will add this example in the tests.

Reply:
OK. I thought that special case would not cover things like `Real[1.0f0]`. Then I guess it's fine, but maybe rename `collection_eltype` to `typeinfo_eltype` to be clear.

I think you can also compress the comments for that function. I love comments, but let's be reasonable. :-) Among other things, there's no need to mention possible alternatives in comments: discussion should happen here, not inside the code base. This also includes the comment above about `isleaftype`.

Reply:
OK for `typeinfo_eltype` and for trying to compress comments! (While writing them when the code is fresh in your head, it's hard to decide what would be a balanced level of verbosity in comments)